sstables/sstable_set.hh

```diff
@@ -99,8 +106,63 @@ public:
         selection select(const dht::ring_position_view& pos) const;
     };
     incremental_selector make_incremental_selector() const;
+
+    flat_mutation_reader create_single_key_sstable_reader(
+        column_family*,
+        schema_ptr,
+        reader_permit,
+        utils::estimated_histogram&,
+        const dht::partition_range&, // must be singular
+        const query::partition_slice&,
+        const io_priority_class&,
+        tracing::trace_state_ptr,
+        streamed_mutation::forwarding,
+        mutation_reader::forwarding) const;
+
+    /// Read a range from the sstable set.
+    ///
+    /// The reader is unrestricted, but will account its resource usage on the
+    /// semaphore belonging to the passed-in permit.
+    flat_mutation_reader make_range_sstable_reader(
+        schema_ptr,
+        reader_permit,
+        const dht::partition_range&,
+        const query::partition_slice&,
+        const io_priority_class&,
+        tracing::trace_state_ptr,
+        streamed_mutation::forwarding,
+        mutation_reader::forwarding,
+        read_monitor_generator& rmg = default_read_monitor_generator()) const;
+
+    // Filters out mutations that don't belong to the current shard.
+    flat_mutation_reader make_local_shard_sstable_reader(
+        schema_ptr,
+        reader_permit,
+        const dht::partition_range&,
+        const query::partition_slice&,
+        const io_priority_class&,
+        tracing::trace_state_ptr,
+        streamed_mutation::forwarding,
+        mutation_reader::forwarding,
+        read_monitor_generator& rmg = default_read_monitor_generator()) const;
 };
 
+/// Read a range from the passed-in sstables.
+///
+/// The reader is restricted, that is it will wait for admission on the semaphore
+/// belonging to the passed-in permit, before starting to read.
+flat_mutation_reader make_restricted_range_sstable_reader(
```

Comment:
Unrelated: all these different overloads are so confusing, we should unify them into one. We could make all reads go through the restricted reader, but bypass it if we determine we can admit the reader straight away. The purpose of the restricted reader is to reduce the memory usage of queued reads, and this is moot if the reader can be admitted immediately on creation.